Cannot open executable Zip files.

Looks like Spring creates [executable jars](https://docs.spring.io/spring-boot/docs/current/reference/html/executable-jar.html) which Okio fails to open.

```
Caused by: java.io.IOException: not a zip: expected 0x4034b50 but was 0x622f2123
	at okio.internal.ZipKt.openZip(zip.kt:75) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem$Companion.toJarRoot(ResourceFileSystem.kt:201) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem$Companion.toClasspathRoots(ResourceFileSystem.kt:184) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem$roots$2.invoke(ResourceFileSystem.kt:47) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem$roots$2.invoke(ResourceFileSystem.kt:47) ~[okio-jvm-3.0.0.jar!/:na]
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74) ~[kotlin-stdlib-1.4.30.jar!/:1.4.30-release-302 (1.4.30)]
	at okio.internal.ResourceFileSystem.getRoots(ResourceFileSystem.kt:47) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem.source(ResourceFileSystem.kt:126) ~[okio-jvm-3.0.0.jar!/:na]
	at com.squareup.wire.schema.CoreLoader.load(CoreLoader.kt:79) ~[wire-schema-jvm-4.1.0.jar!/:na]
```

See https://github.com/square/wire/discussions/2209

Hi - I'm the OP on the original issue.

I've tried a couple of things on this one, but haven't found any success.

## Made the spring boot jar non executable.
```
<plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <configuration>
                    <executable>false</executable>
                </configuration>
```

This at least removes the `not a zip:` error.
However, I quickly fall at the next hurdle.

## Export the dependencies explicitly
Turns out that the Spring Boot plugin will allow you to make [specific jars explicitly available](https://docs.spring.io/spring-boot/docs/1.4.x/reference/htmlsingle/#howto-extract-specific-libraries-when-an-executable-jar-runs) in the fat jar. 

This kinda worked - for the JARs I explicitly exported, it looks like the ResourceFileSystem loaded them successfully.  But it looks CoreLoader is iterating over every jar in the classpath, and throws NullPointer exceptions when the file isn't available.  

Given `ResourceFileSystem` declares the following:

```
fun URL.toJarRoot(): Pair<FileSystem, Path>? {
      val urlString = toString()
      if (!urlString.startsWith("jar:file:")) return null // Ignore unexpected URLs.
...
```

A possible alternative is that if the file is not found, to log and return null.  This would at least allow loading to continue.

## Use different Spring Boot layouts
Spring boot let us change the [layout](https://docs.spring.io/spring-boot/docs/current/maven-plugin/reference/htmlsingle/#goals-repackage-parameters-details-layout).

I've tried all the options, and all threw FileNotFound exceptions:

```
Caused by: java.io.FileNotFoundException: /home/martypitt/dev/vyne-demos/films/films-api/target/films-api.jar!/BOOT-INF/lib/films-kafka-publisher-0.1.0-SNAPSHOT.jar (No such file or directory)
	at java.base/java.io.RandomAccessFile.open0(Native Method) ~[na:na]
	at java.base/java.io.RandomAccessFile.open(RandomAccessFile.java:345) ~[na:na]
	at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:259) ~[na:na]
	at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:214) ~[na:na]
	at okio.JvmSystemFileSystem.openReadOnly(JvmSystemFileSystem.kt:82) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ZipKt.openZip(zip.kt:67) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem$Companion.toJarRoot(ResourceFileSystem.kt:201) ~[okio-jvm-3.0.0.jar!/:na]
	at okio.internal.ResourceFileSystem$Companion.toClasspathRoots(ResourceFileSystem.kt:184) ~[okio-jvm-3.0.0.jar!/:na]
```

So, still a bit stuck, but thought I'd share the progress / research.

--- a/gradle.properties
+++ b/gradle.properties
@@ -2,6 +2,7 @@ org.gradle.jvmargs='-Dfile.encoding=UTF-8'
 android.enableJetifier=true
 android.useAndroidX=true
 kotlin.js.compiler=both
+kotlin.mpp.stability.nowarn=true
 
 # Publishing SHA 256 and 512 hashses of maven-metadata is not supported by Sonatype and Nexus.
 # See https://github.com/gradle/gradle/issues/11308 and

--- a/okio/src/jvmMain/kotlin/okio/internal/zip.kt
+++ b/okio/src/jvmMain/kotlin/okio/internal/zip.kt
@@ -65,19 +65,6 @@ internal fun openZip(
   predicate: (ZipEntry) -> Boolean = { true }
 ): ZipFileSystem {
   fileSystem.openReadOnly(zipPath).use { fileHandle ->
-
-    fileHandle.source().buffer().use { source ->
-      val firstFileSignature = source.readIntLe()
-      if (firstFileSignature != LOCAL_FILE_HEADER_SIGNATURE) {
-        if (firstFileSignature == END_OF_CENTRAL_DIRECTORY_SIGNATURE) {
-          throw IOException("unsupported zip: empty")
-        }
-        throw IOException(
-          "not a zip: expected ${LOCAL_FILE_HEADER_SIGNATURE.hex} but was ${firstFileSignature.hex}"
-        )
-      }
-    }
-
     // Scan backwards from the end of the file looking for the END_OF_CENTRAL_DIRECTORY_SIGNATURE.
     // If this file has no comment we'll see it on the first attempt; otherwise we have to go
     // backwards byte-by-byte until we reach it. (The number of bytes scanned will equal the comment
@@ -132,6 +119,10 @@ internal fun openZip(
           }
         }
       }
+    }
+
+    if (record.entryCount == 0L) {
+      throw IOException("unsupported zip: empty")
     }
 
     // Seek to the first central directory entry and read all of the entries.

--- a/okio/src/jvmTest/java/okio/ZipFileSystemTest.kt
+++ b/okio/src/jvmTest/java/okio/ZipFileSystemTest.kt
@@ -16,6 +16,7 @@
 package okio
 
 import kotlinx.datetime.Instant
+import okio.ByteString.Companion.decodeHex
 import okio.ByteString.Companion.encodeUtf8
 import okio.Path.Companion.toPath
 import org.assertj.core.api.Assertions.assertThat
@@ -30,6 +31,35 @@ class ZipFileSystemTest {
   @Before
   fun setUp() {
     fileSystem.createDirectory(base)
+  }
+
+  @Test
+  fun emptyZipThrows() {
+    // ZipBuilder cannot write empty zips.
+    val zipPath = base / "empty.zip"
+    fileSystem.write(zipPath) {
+      write("504b0506000000000000000000000000000000000000".decodeHex())
+    }
+
+    val t = assertFailsWith<IOException> {
+      fileSystem.openZip(zipPath)
+    }
+    assertThat(t).hasMessage("unsupported zip: empty")
+  }
+
+  @Test
+  fun emptyZipWithPrependedDataThrows() {
+    // ZipBuilder cannot write empty zips.
+    val zipPath = base / "empty.zip"
+    fileSystem.write(zipPath) {
+      writeUtf8("Hello I'm junk data prepended to the ZIP!")
+      write("504b0506000000000000000000000000000000000000".decodeHex())
+    }
+
+    val t = assertFailsWith<IOException> {
+      fileSystem.openZip(zipPath)
+    }
+    assertThat(t).hasMessage("unsupported zip: empty")
   }
 
   @Test

